odl/discr/discr_utils.py

```diff
@@ -0,0 +1,812 @@
+    See the `ODL vectorization guide`_ for a detailed introduction.
+    See Also
+    --------
+    make_func_for_sampling : wrap a function
+    odl.discr.grid.RectGrid.meshgrid
+    numpy.meshgrid
+    References
+    ----------
+    .. _ODL vectorization guide:
+       https://odlgroup.github.io/odl/guide/in_depth/vectorization_guide.html
+    """
+    if out is None:
+        out = func(points, **kwargs)
+    else:
+        func(points, out=out, **kwargs)
+    return out
+
+
+def nearest_interpolator(f, coord_vecs, variant='left'):
+    """Return the nearest neighbor interpolator for discrete values.
+    Given points ``x[1] < x[2] < ... < x[N]``, and function values
+    ``f[1], ..., f[N]``, nearest neighbor interpolation at ``x`` is defined
+    as ::
+        I(x) = f[j]  with j such that |x - x[j]| is minimal.
+    The ambiguity at the midpoints is resolved by preferring one of the
+    neighbors. In higher dimensions, this principle is applied per axis.
+    The returned interpolator is the piecewise constant function ``x -> I(x)``.
+    Parameters
+    ----------
+    f : numpy.ndarray
+        Function values that should be interpolated.
+    coord_vecs : sequence of numpy.ndarray
+        Coordinate vectors of the rectangular grid on which interpolation
+        should be based. They must be sorted in ascending order. Usually
+        they are obtained as ``grid.coord_vectors`` from a `RectGrid`.
+    variant : {'left', 'right'}, optional
+        Which neighbor to prefer at the midpoint between two nodes.
+    Returns
+    -------
+    interpolator : function
+        Python function that will interpolate the given values when called
+        with a point or multiple points (vectorized).
+    Examples
+    --------
+    We interpolate a 1d function. If called with a single point, the
+    interpolator returns a single value, and with multiple points at once,
+    an array of values is returned:
+    >>> part = odl.uniform_partition(0, 2, 5)
+    >>> part.coord_vectors  # grid points
+    (array([ 0.2,  0.6,  1. ,  1.4,  1.8]),)
+    >>> f = [1, 2, 3, 4, 5]
+    >>> interpolator = nearest_interpolator(f, part.coord_vectors)
+    >>> interpolator(0.3)  # closest to 0.2 -> value 1
+    1
+    >>> interpolator([0.6, 1.3, 1.9])  # closest to [0.6, 1.4, 1.8]
+    array([2, 4, 5])
+    In 2 dimensions, we can either use a list of points or a meshgrid:
+    >>> part = odl.uniform_partition([0, 0], [1, 5], shape=(2, 4))
+    >>> part.coord_vectors  # grid points
+    (array([ 0.25,  0.75]), array([ 0.625,  1.875,  3.125,  4.375]))
+    >>> f = np.array([[1, 2, 3, 4],
+    ...               [5, 6, 7, 8]],
+    ...              dtype=float)
+    >>> interpolator = nearest_interpolator(f, part.coord_vectors)
+    >>> interpolator([1, 1])  # single point
+    5.0
+    >>> interpolator([[0.5, 2.0],
+    ...               [0.0, 4.5],
+    ...               [0.0, 3.0]])  # 3 points at once
+    array([ 4.,  5.,  3.])
```

Comment:
I am not sure I understand how this interpolator works in higher dimensions... why is this not ``array([ 2.,  4.,  3.])``?

Reply:
Don't take the doctests too seriously at the moment. Need fix anyway. (Means: probably it should be `[2, 4, 3]`).